File: frontend/src/services/api/groups.ts
import apiClient from './client';

// 그룹 API 타입 정의
export interface GroupOwner {
  user_id: number;
  nickname: string;
  profile_image_url?: string | null;
  selected_achievement?: {
    achievement_id: number;
    title: string;
    description?: string | null;
    icon_url?: string | null;
  } | null;
}

export interface GroupMember {
  member_id: number;
  user_id: number;
  nickname: string;
  profile_image_url?: string | null;
  selected_achievement?: {
    achievement_id: number;
    title: string;
    description?: string | null;
    icon_url?: string | null;
  } | null;
  role: 'owner' | 'admin' | 'member';
  joined_at: string;
}

export interface Group {
  group_id: number;
  group_name: string;
  description?: string | null;
  owner_id: number;
  owner: GroupOwner;
  is_public: boolean;
  max_members: number;
  member_count: number;
  current_user_role?: 'owner' | 'admin' | 'member' | null;
  is_member: boolean;
  created_at: string;
}

export interface GroupListResponse {
  groups: Group[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface GroupMemberListResponse {
  members: GroupMember[];
  total: number;
}

export interface GroupCreate {
  group_name: string;
  description?: string;
  is_public?: boolean;
  max_members?: number;
}

export interface GroupUpdate {
  group_name?: string;
  description?: string;
  is_public?: boolean;
  max_members?: number;
}

export interface MessageResponse {
  message: string;
}

export interface GroupInvitation {
  invitation_id: number;
  group_id: number;
  group: {
    group_id: number;
    group_name: string;
    description?: string | null;
    is_public: boolean;
  };
  inviter_id: number;
  inviter: {
    user_id: number;
    nickname: string;
    profile_image_url?: string | null;
    selected_achievement?: {
      achievement_id: number;
      title: string;
      description?: string | null;
      icon_url?: string | null;
    } | null;
  };
  invitee_id: number;
  invitee: {
    user_id: number;
    nickname: string;
    profile_image_url?: string | null;
    selected_achievement?: {
      achievement_id: number;
      title: string;
      description?: string | null;
      icon_url?: string | null;
    } | null;
  };
  status: 'pending' | 'accepted' | 'declined' | 'expired';
  created_at: string;
  updated_at: string;
}

export interface GroupInvitationListResponse {
  invitations: GroupInvitation[];
  total: number;
}

export interface GroupInvitationCreate {
  invitee_id: number;
}

export interface GroupMemberStatistics {
  user_id: number;
  nickname: string;
  profile_image_url?: string | null;
  total_practice_time: number;
  total_sessions: number;
  consecutive_days: number;
  last_practice_date?: string | null;
  average_session_time?: number | null;
}

export interface GroupStatistics {
  group_id: number;
  total_members: number;
  total_practice_time: number;
  total_sessions: number;
  average_practice_time_per_member: number;
  average_sessions_per_member: number;
  most_active_member?: GroupMemberStatistics | null;
  weekly_practice_data: number[];
}

export interface GroupMemberStatisticsListResponse {
  members: GroupMemberStatistics[];
  total: number;
}

export const groupsApi = {
  /**
   * 그룹 목록 조회
   */
  getGroups: async (params?: {
    page?: number;
    page_size?: number;
    is_public?: boolean;
    search?: string;
  }): Promise<GroupListResponse> => {
    const response = await apiClient.get<GroupListResponse>('/groups', { params });
    return response.data;
  },

  /**
   * 그룹 생성
   */
  createGroup: async (data: GroupCreate): Promise<Group> => {
    const response = await apiClient.post<Group>('/groups', data);
    return response.data;
  },

  /**
   * 그룹 상세 조회
   */
  getGroup: async (group_id: number): Promise<Group> => {
    const response = await apiClient.get<Group>(`/groups/${group_id}`);
    return response.data;
  },

  /**
   * 그룹 가입
   */
  joinGroup: async (group_id: number): Promise<MessageResponse> => {
    const response = await apiClient.post<MessageResponse>(`/groups/${group_id}/join`);
    return response.data;
  },

  /**
   * 그룹 탈퇴
   */
  leaveGroup: async (group_id: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`/groups/${group_id}/leave`);
    return response.data;
  },

  /**
   * 그룹 멤버 목록 조회
   */
  getGroupMembers: async (group_id: number): Promise<GroupMemberListResponse> => {
    const response = await apiClient.get<GroupMemberListResponse>(`/groups/${group_id}/members`);
    return response.data;
  },

  /**
   * 그룹 정보 수정
   */
  updateGroup: async (group_id: number, data: GroupUpdate): Promise<Group> => {
    const response = await apiClient.put<Group>(`/groups/${group_id}`, data);
    return response.data;
  },

  /**
   * 그룹 삭제
   */
  deleteGroup: async (group_id: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`/groups/${group_id}`);
    return response.data;
  },

  /**
   * 그룹 초대 보내기
   */
  inviteMember: async (group_id: number, data: GroupInvitationCreate): Promise<GroupInvitation> => {
    const response = await apiClient.post<GroupInvitation>(`/groups/${group_id}/invitations`, data);
    return response.data;
  },

  /**
   * 내가 받은 그룹 초대 목록 조회
   */
  getGroupInvitations: async (params?: {
    status?: 'pending' | 'accepted' | 'declined' | 'expired';
  }): Promise<GroupInvitationListResponse> => {
    // params가 undefined이거나 빈 객체인 경우를 처리
    const queryParams: Record<string, string> = {};
    if (params?.status) {
      queryParams.status = params.status;
    }
    const response = await apiClient.get<GroupInvitationListResponse>('/groups/invitations', { 
      params: Object.keys(queryParams).length > 0 ? queryParams : undefined 
    });
    return response.data;
  },

  /**
   * 그룹 초대 수락
   */
  acceptInvitation: async (invitation_id: number): Promise<MessageResponse> => {
    const response = await apiClient.post<MessageResponse>(`/groups/invitations/${invitation_id}/accept`);
    return response.data;
  },

  /**
   * 그룹 초대 거절
   */
  declineInvitation: async (invitation_id: number): Promise<MessageResponse> => {
    const response = await apiClient.post<MessageResponse>(`/groups/invitations/${invitation_id}/decline`);
    return response.data;
  },

  /**
   * 그룹 전체 통계 조회
   */
  getGroupStatistics: async (group_id: number): Promise<GroupStatistics> => {
    const response = await apiClient.get<GroupStatistics>(`/groups/${group_id}/statistics`);
    return response.data;
  },

  /**
   * 그룹 멤버별 통계 조회
   */
  getGroupMemberStatistics: async (group_id: number): Promise<GroupMemberStatisticsListResponse> => {
    const response = await apiClient.get<GroupMemberStatisticsListResponse>(`/groups/${group_id}/members/statistics`);
    return response.data;
  },
};

File: backend/app/schemas/groups.py
"""
그룹 관련 Pydantic 스키마
그룹 생성, 조회, 멤버 관리 등의 스키마 정의
"""
from pydantic import BaseModel, Field
from typing import Optional, List, TYPE_CHECKING
from datetime import datetime

if TYPE_CHECKING:
    from app.schemas.achievements import AchievementResponse


# ========== 그룹 스키마 ==========

class GroupCreate(BaseModel):
    """그룹 생성 요청 스키마"""
    group_name: str = Field(..., min_length=1, max_length=200, description="그룹 이름")
    description: Optional[str] = Field(None, description="그룹 설명")
    is_public: bool = Field(default=False, description="공개 그룹 여부")
    max_members: int = Field(default=50, ge=1, le=1000, description="최대 멤버 수")

    class Config:
        json_schema_extra = {
            "example": {
                "group_name": "피아노 연습 모임",
                "description": "피아노를 연주하는 분들을 위한 그룹입니다.",
                "is_public": True,
                "max_members": 50
            }
        }


class GroupUpdate(BaseModel):
    """그룹 수정 요청 스키마"""
    group_name: Optional[str] = Field(None, min_length=1, max_length=200, description="그룹 이름")
    description: Optional[str] = Field(None, description="그룹 설명")
    is_public: Optional[bool] = Field(None, description="공개 그룹 여부")
    max_members: Optional[int] = Field(None, ge=1, le=1000, description="최대 멤버 수")

    class Config:
        json_schema_extra = {
            "example": {
                "group_name": "수정된 그룹 이름",
                "description": "수정된 그룹 설명",
                "is_public": False,
                "max_members": 100
            }
        }


class GroupOwnerResponse(BaseModel):
    """그룹 소유자 정보 응답 스키마"""
    user_id: int
    nickname: str
    profile_image_url: Optional[str] = None
    selected_achievement: Optional["AchievementResponse"] = None  # 선택한 칭호 정보

    class Config:
        from_attributes = True


class GroupMemberResponse(BaseModel):
    """그룹 멤버 정보 응답 스키마"""
    member_id: int
    user_id: int
    nickname: str
    profile_image_url: Optional[str] = None
    selected_achievement: Optional["AchievementResponse"] = None  # 선택한 칭호 정보
    role: str  # 'owner', 'admin', 'member'
    joined_at: datetime

    class Config:
        from_attributes = True


class GroupResponse(BaseModel):
    """그룹 응답 스키마"""
    group_id: int
    group_name: str
    description: Optional[str] = None
    owner_id: int
    owner: GroupOwnerResponse
    is_public: bool
    max_members: int
    member_count: int  # 현재 멤버 수
    current_user_role: Optional[str] = None  # 현재 사용자의 역할 (가입하지 않았으면 None)
    is_member: bool = False  # 현재 사용자가 멤버인지 여부
    created_at: datetime

    class Config:
        from_attributes = True


class GroupListResponse(BaseModel):
    """그룹 목록 응답 스키마"""
    groups: List[GroupResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

    class Config:
        json_schema_extra = {
            "example": {
                "groups": [],
                "total": 0,
                "page": 1,
                "page_size": 20,
                "total_pages": 0
            }
        }


class GroupMemberListResponse(BaseModel):
    """그룹 멤버 목록 응답 스키마"""
    members: List[GroupMemberResponse]
    total: int

    class Config:
        json_schema_extra = {
            "example": {
                "members": [],
                "total": 0
            }
        }


class MessageResponse(BaseModel):
    """일반 메시지 응답 스키마"""
    message: str

    class Config:
        json_schema_extra = {
            "example": {
                "message": "작업이 성공적으로 완료되었습니다."
            }
        }


# ========== 그룹 초대 스키마 ==========

class GroupInvitationCreate(BaseModel):
    """그룹 초대 생성 요청 스키마"""
    invitee_id: int = Field(..., description="초대받을 사용자 ID")

    class Config:
        json_schema_extra = {
            "example": {
                "invitee_id": 2
            }
        }


class GroupInvitationInviterResponse(BaseModel):
    """초대자 정보 응답 스키마"""
    user_id: int
    nickname: str
    profile_image_url: Optional[str] = None
    selected_achievement: Optional["AchievementResponse"] = None  # 선택한 칭호 정보

    class Config:
        from_attributes = True


class GroupInvitationInviteeResponse(BaseModel):
    """초대받은 사용자 정보 응답 스키마"""
    user_id: int
    nickname: str
    profile_image_url: Optional[str] = None
    selected_achievement: Optional["AchievementResponse"] = None  # 선택한 칭호 정보

    class Config:
        from_attributes = True


class GroupInvitationGroupResponse(BaseModel):
    """초대된 그룹 정보 응답 스키마"""
    group_id: int
    group_name: str
    description: Optional[str] = None
    is_public: bool

    class Config:
        from_attributes = True


class GroupInvitationResponse(BaseModel):
    """그룹 초대 응답 스키마"""
    invitation_id: int
    group_id: int
    group: GroupInvitationGroupResponse
    inviter_id: int
    inviter: GroupInvitationInviterResponse
    invitee_id: int
    invitee: GroupInvitationInviteeResponse
    status: str  # 'pending', 'accepted', 'declined', 'expired'
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GroupInvitationListResponse(BaseModel):
    """그룹 초대 목록 응답 스키마"""
    invitations: List[GroupInvitationResponse]
    total: int

    class Config:
        json_schema_extra = {
            "example": {
                "invitations": [],
                "total": 0
            }
        }

